lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3970,21 +3960,41 @@ DifferentiationTask::DifferentiationTask(ADContext &context,
                                          DifferentiationInvoker invoker)
     : context(context), original(original), attr(attr), invoker(invoker) {
   auto &module = context.getModule();
+  // If attribute specifies JVP name, try to look up JVP in current module.
+  // If the JVP doesn't exist, or if the original is an external declaration,
+  // create an external JVP reference.
   if (attr->hasJVP()) {
-    // If attribute specifies JVP name, try to look up JVP in current module.
-    // Otherwise, create an external reference.
     jvp = module.lookUpFunction(attr->getJVPName());
     if (!jvp)
       jvp = context.declareExternalAssociatedFunction(
-          original, attr, AutoDiffAssociatedFunctionKind::JVP);
-  }
+          original, attr, attr->getJVPName(),
+          AutoDiffAssociatedFunctionKind::JVP);
+  } else if (original->isExternalDeclaration()) {
+    auto jvpName = original->getASTContext()
+                       .getIdentifier("AD__" + original->getName().str() +
+                                      "__jvp_" + getIndices().mangle())
+                       .str();
+    attr->setJVPName(jvpName);
```

Comment:
Move the set-name logic to `declareExternalAssociatedFunction`.

Reply:
Done in https://github.com/apple/swift/pull/22280/commits/6e8f9d503b12611a367527a41b1e98c0ef3ec9d2.